Refactor image management to use transfer service

ChengyuZhu6 opened the pull request:
Fixes: #4573

ChengyuZhu6 committed:
configure containerd config with remote snapshotter

Add remote snapshot annotations and transfer unpack config
for stargz, soci, and fuse-overlayfs snapshotter plugins.

Signed-off-by: ChengyuZhu6 <hudson@cyzhu.com>

ChengyuZhu6 committed:
Refactor image management to use transfer service

Switch image operations to the transfer API with structured progress reporting
and improved TLS/HTTP fallback behavior. Introduce shared helpers for
credentials, error classification, progress rendering, and transfer-based
import/tag/save flows, updating tests to reflect the new UX.

Signed-off-by: ChengyuZhu6 <hudson@cyzhu.com>

AkihiroSuda commented on cmd/nerdctl/image/image_import.go:
```diff
@@ -45,6 +45,7 @@ func ImportCommand() *cobra.Command {
 
 	cmd.Flags().StringP("message", "m", "", "Set commit message for imported image")
 	cmd.Flags().String("platform", "", "Set platform for imported image (e.g., linux/amd64)")
+	cmd.Flags().Bool("local", false, "Import content locally rather than using transfer service")
```
For simplicity can we just remove the support for local mode?

ChengyuZhu6 replied:
Ok, but this means that containerd versions <= 1.6 will be unsupported starting with the next release.

ChengyuZhu6 committed:
Support legacy push/pull for containerd 1.7.x

Add version detection to automatically select Transfer service (2.0+)
or legacy resolver methods (< 2.0) for better compatibility.

Signed-off-by: ChengyuZhu6 <hudson@cyzhu.com>

AkihiroSuda commented on Dockerfile.d/etc_containerd_config.toml:
```diff
@@ -5,3 +5,12 @@ version = 2
+    enable_remote_snapshot_annotations = "true"
+[[plugins."io.containerd.transfer.v1.local".unpack_config]]
+  platform = "linux"
+  snapshotter = "overlayfs"
```
What happens if this config is not specified?
i.e., what happens with the default config?

AkihiroSuda commented on pkg/containerdutil/version.go:
```diff
@@ -0,0 +1,50 @@
+}
+
+// SupportsFullTransferService checks if the containerd version fully supports the Transfer service.
+// While containerd 1.7 has Transfer service, full support is only available in 2.0+.
```
```suggestion
// While containerd 1.7 has Transfer service, full support is only available in 2.0+.
// The following features are missing in containerd 1.7:
// - [ADD MISSING FEATURES]
```

ChengyuZhu6 replied:
Done.

AkihiroSuda commented on extras/rootless/containerd-rootless-setuptool.sh:
```diff
@@ -404,6 +404,15 @@ cmd_entrypoint_install_fuse_overlayfs() {
+			snapshotter = "fuse-overlayfs"
+		[[plugins."io.containerd.transfer.v1.local".unpack_config]]
+			platform = "linux"
+			snapshotter = "overlayfs"
```
Not a fan of complicating the config.
What happens if these lines are missing?
ctr doesn't need them AFAICS?

ChengyuZhu6 replied:
Sure.

AkihiroSuda reviewed (approved):
Thanks
